api/grpc_server.go

```diff
@@ -108,13 +121,17 @@ func (s SpacemeshGrpcService) GetBalance(ctx context.Context, in *pb.AccountId)
 		return nil, fmt.Errorf("account does not exist")
 	}
 
-	val := s.StateApi.GetBalance(addr)
-	valStr := strconv.FormatUint(val, 10)
-	log.Info("GRPC GetBalance returned msg %v", valStr)
-	return &pb.SimpleMessage{Value: valStr}, nil
+	_, balance, err := s.getProjection(addr)
+	if err != nil {
+		return nil, err
+	}
+	msg := &pb.SimpleMessage{Value: strconv.FormatUint(balance, 10)}
+	log.Info("GRPC GetBalance returned msg.Value %v", msg.Value)
+	return msg, nil
 }
 
-// Echo returns the response for an echo api request
+// GetNonce returns the current account nonce for the provided account ID. The nonce is based on the global state and
+// all known transactions in unapplied blocks and the mempool.
 func (s SpacemeshGrpcService) GetNonce(ctx context.Context, in *pb.AccountId) (*pb.SimpleMessage, error) {
```

Comment:
I think these funcitons should return 2 values, 1st is the projected state, and the 2nd is the actual state. since projected state isn't confirmed I don't think that the wallet should rely on it to display balance for example

Reply:
I agree.

AFAIK the wallet doesn't display the nonce, but only uses it internally to create transactions, so it's only relevant to balance.

I think that we should display two values: "Balance" and "Available Balance".

"Balance" is the actual confirmed balance from the global state, while "Available Balance" is what the wallet will allow you to spend, which takes pending transactions into account.

This is actually not a new concept, I know it from working in the FinTech industry and most US banks show these two kinds of balance to their customers. One is what's in your account and the other considers expected movements that haven't been applied to the account yet.

Having said all this, I think that we can live with only the projected (lower) amount for now, and consider this improvement for later.

cc: @avive